include/irr/asset/ICPUVirtualTexture.h

```diff
@@ -0,0 +1,337 @@
+                    if (x == w-1u)
+                        copy.extentLayerCount.x = std::max(extent.width>>(_subres.baseMipLevel+i),1u)-copy.inOffsetBaseLayer.x;
+                    if (y == h-1u)
+                        copy.extentLayerCount.y = std::max(extent.height>>(_subres.baseMipLevel+i),1u)-copy.inOffsetBaseLayer.y;
+                    memcpy(&copy.paddedExtent.width,(copy.extentLayerCount+core::vectorSIMDu32(2u*m_tilePadding)).pointer, 2u*sizeof(uint32_t));
+                    copy.paddedExtent.depth = 1u;
+                    if (w>1u)
+                        copy.extentLayerCount.x += m_tilePadding;
+                    if (x>0u && x<w-1u)
+                        copy.extentLayerCount.x += m_tilePadding;
+                    if (h>1u)
+                        copy.extentLayerCount.y += m_tilePadding;
+                    if (y>0u && y<h-1u)
+                        copy.extentLayerCount.y += m_tilePadding;
+                    if (x == 0u)
+                        copy.relativeOffset.x = m_tilePadding;
+                    else
+                        copy.inOffsetBaseLayer.x -= m_tilePadding;
+                    if (y == 0u)
+                        copy.relativeOffset.y = m_tilePadding;
+                    else
+                        copy.inOffsetBaseLayer.y -= m_tilePadding;
+                    copy.inMipLevel = _subres.baseMipLevel + i;
+                    copy.outMipLevel = 0u;
+                    copy.inImage = _img;
+                    copy.outImage = storage->image.get();
+                    copy.axisWraps[0] = _wrapu;
+                    copy.axisWraps[1] = _wrapv;
+                    copy.axisWraps[2] = ISampler::ETC_CLAMP_TO_EDGE;
+                    copy.borderColor = _borderColor;
+                    if (!CPaddedCopyImageFilter::execute(&copy))
+                        _IRR_DEBUG_BREAK_IF(true);
+                }
+        }
+
+        return offsetToTextureData(pgtOffset, _img, _wrapu, _wrapv);
+    }
+
+    bool free(const STextureData& _addr, const IImage* _img, const IImage::SSubresourceRange& _subres) override
+    {
+        const E_FORMAT format = _img->getCreationParameters().format;
+        ICPUVTResidentStorage* storage = nullptr;
+        {
+            auto found = m_storage.find(getFormatClass(format));
+            if (found==m_storage.end())
+                return false;
+            storage = static_cast<ICPUVTResidentStorage*>(found->second.get());
+        }
+
+        //free physical pages
+        auto extent = _img->getCreationParameters().extent;
+        const uint32_t levelCount = countLevelsTakingAtLeastOnePage(extent, _subres);
+
+        CFillImageFilter::state_type fill;
+        fill.outImage = m_pageTable.get();
+        fill.subresource.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+        fill.subresource.baseArrayLayer = _addr.pgTab_layer;
+        fill.subresource.layerCount = 1u;
+        fill.fillValue.asUint.x = SPhysPgOffset::invalid_addr;
```

Comment:
only do the fill in `_IRR_DEBUG` its completely acceptable to have garbage in the page table entries that do not have an allocation (nobody is supposed to query those regions, at all!)

Reply:
plus anyway, you're doing it twice... cause you've already done it here?
https://github.com/buildaworldnet/IrrlichtBAW/pull/474/files?file-filters%5B%5D=.cpp&file-filters%5B%5D=.h&file-filters%5B%5D=No+extension&hide-deleted-files=true#r417913790